model/model_test.go

```diff
@@ -0,0 +1,157 @@
+package model
+
+import (
+	"testing"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/stretchr/testify/assert"
+
+	"github.com/signalfx/com_signalfx_metrics_protobuf"
+)
+
+func TestUnmarshalMarshal(t *testing.T) {
+	model := generateDataPointUploadMessage()
+	bytes, _ := model.Marshal()
+	proto2Model := &com_signalfx_metrics_protobuf.DataPointUploadMessage{}
+	if err := proto.Unmarshal(bytes, proto2Model); err != nil {
+		t.Fail()
+	}
+	var lightBytes []byte
+	var err error
+	if lightBytes, err = proto.Marshal(proto2Model); err != nil {
+		t.Fail()
+	}
+	newModel := &DataPointUploadMessage{}
+	if err := newModel.Unmarshal(lightBytes); err != nil {
+		t.Fail()
+	}
+	assert.EqualValues(t, model, newModel)
+}
+
+func BenchmarkProto2Model(b *testing.B) {
+	uploadMessage := generateDataPointUploadMessage()
+	bytes, _ := uploadMessage.Marshal()
+
+	b.ResetTimer()
+	b.ReportAllocs()
+	for i := 0; i < b.N; i++ {
+		proto2Model := &com_signalfx_metrics_protobuf.DataPointUploadMessage{}
+		if err := proto.Unmarshal(bytes, proto2Model); err != nil {
+			b.Fail()
+		}
+		if _, err := proto.Marshal(proto2Model); err != nil {
+			b.Fail()
+		}
+	}
+}
+
+func BenchmarkModel(b *testing.B) {
+	uploadMessage := generateDataPointUploadMessage()
+	bytes, _ := uploadMessage.Marshal()
+
+	b.ResetTimer()
+	b.ReportAllocs()
+	for i := 0; i < b.N; i++ {
+		model := &DataPointUploadMessage{}
+		if err := model.Unmarshal(bytes); err != nil {
+			b.Fail()
+		}
+		if _, err := model.Marshal(); err != nil {
+			b.Fail()
+		}
+	}
+}
+
+func generateDataPointUploadMessage() *DataPointUploadMessage {
+	ret := &DataPointUploadMessage{}
+	ret.Datapoints = append(ret.Datapoints, generateEnumMetrics(10)...)
```

Comment:
maybe combine 3 api's generateEnumMetrics, generateGaugeMetrics, generateCounterMetrics into one `generateMetrics(numPoints, metricType)`, inside the generate api
` var d Datum
   switch metricType {
case Gauge:
       d.DoubleValue = pointer of 12.3
}
` 
similarly for other cases

Reply:
I like more explicit code in tests, to make it easier to read. I agree if this would have been non-test code your suggestion would be the way to go.